### ISSUE TITLE
Tools integrating semgrep  programatically breaks as the API is being updated on releases.

### ISSUE THREAD

**Issue by ajinabraham** [semgrep/semgrep#2880]
**Is your feature request related to a problem? Please describe.**
njsscan was using semgrep 0.38 and a recent upgrade to 0.45 breaks tests

```
        output_handler = OutputHandler(
>           OutputSettings(
                output_format=OutputFormat.JSON,
                output_destination=None,
                error_on_findings=False,
                verbose_errors=False,
                strict=False,
                timeout_threshold=3,
                json_stats=False,
                output_per_finding_max_lines_limit=None,
            ),
            stdout=io_capture,
        )
E       TypeError: __new__() missing 1 required positional argument: 'json_time'

```

Current njsscan integration with semgrep: https://github.com/ajinabraham/libsast/blob/master/libsast/core_sgrep/helpers.py#L9

**Describe the solution you'd like**

Can we have two sets of APIs?

For seamless integration and maintenance, the semgrep API has to follow a contract with third-party tools. I suggest an internal and external API. The internal API changes and evolves with semgrep and the external one is fixed, documented, and changes only on major releases / a major security update.

**Describe alternatives you've considered**
NA

**Additional context**
NA

**Comment by mschwager** [comment]
cc @mjambon who added the `json_time` functionality (https://github.com/returntocorp/semgrep/pull/2702). Is there a reason this isn't optional and defaulting to `False`?

@ajinabraham you can fix this by adding a `json_time=False` line to your `OutputSettings` call :+1:

**Comment by ajinabraham** [comment]
I fixed it by adding `json_time`. @mschwager, it would be nice to have a stable API for programmatic integration with semgrep.

**Comment by aryx** [comment]
@mschwager what about we have a semgrep_api_v1.py somewhere, with big comments at the top we can't change the signatures of the functions?

**Comment by mschwager** [comment]
> @mschwager what about we have a semgrep_api_v1.py somewhere, with big comments at the top we can't change the signatures of the functions?

I like it, that sounds good to me :+1:

**Comment by aryx** [comment]
For reference, I'm doing something similar on the generic AST side: https://github.com/returntocorp/semgrep/pull/2891
with a generic_ast_v1.atd

**Comment by brendongo** [comment]
I think if we have https://github.com/returntocorp/semgrep/blob/develop/semgrep/semgrep/semgrep_main.py#L132 actually set fields based on arguments it would work as an API. This function is run in tests intending it to be the entry point for calling from other python files

**Comment by mschwager** [comment]
> This function is run in tests intending it to be the entry point for calling from other python files

It looks like the issue here was the `OutputSettings` interface changed, not `invoke_semgrep`. Since we're already taking `**kwargs` and passing them to `main`, it would be difficult to take additional `kwargs` for `OutputSettings`. I can think of a couple options to fix this:

- We have `invoke_semgrep` take an optional `output_settings_kwargs` argument and pass that to `OutputSettings`
- We have `invoke_semgrep` take an optional `output_settings` argument and use dependency injection
  - Basically, if `output_settings` is passed from the caller use that, else use our defaults currently set

Then in the `OutputSettings` class definition we can add a note like "additions here should be backwards compatible to maintain API compatibility." What do you think?

**Comment by brendongo** [comment]
IMO we should make invoke_semgrep more useful as an API in letting people pass settings to it (right now I think a lot of options are hardcoded) then setting sane default values to invoke_semgrep whenever new options are added to OutputSettings should make it pretty stable
